include/circt/Support/ConversionPatternSet.h

```diff
@@ -0,0 +1,69 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_UNUSEDOPPRUNER_H
+#define CIRCT_SUPPORT_UNUSEDOPPRUNER_H
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+
+/// Utility class to simplify populating a `RewritePatternSet` with conversion
+/// patterns.
+struct ConversionPatternSet {
+  RewritePatternSet &patterns;
+  const TypeConverter &typeConverter;
+
+  ConversionPatternSet(RewritePatternSet &patterns,
+                       const TypeConverter &typeConverter)
+      : patterns(patterns), typeConverter(typeConverter) {}
+
+  /// Add multiple conversion patterns to the set.
+  template <typename... Ts>
+  ConversionPatternSet &add() {
+    patterns.add<Ts...>(typeConverter, patterns.getContext());
+    return *this;
+  }
+
+  /// Add a `matchAndRewrite` function as a conversion pattern to the set.
+  template <class Op>
+  ConversionPatternSet &add(LogicalResult (*implFn)(Op, PatternRewriter &)) {
+    patterns.add(implFn);
+    return *this;
+  }
+
+  /// Add a `matchAndRewrite` function as a conversion pattern to the set.
+  template <class Op>
+  ConversionPatternSet &
+  add(LogicalResult (*implFn)(Op, typename Op::Adaptor,
+                              ConversionPatternRewriter &)) {
+
+    struct FnPattern final : public OpConversionPattern<Op> {
+      using OpConversionPattern<Op>::OpConversionPattern;
+      LogicalResult (*implFn)(Op, typename Op::Adaptor,
+                              ConversionPatternRewriter &);
+
+      LogicalResult
+      matchAndRewrite(Op op, typename Op::Adaptor adaptor,
+                      ConversionPatternRewriter &rewriter) const override {
+        return implFn(op, adaptor, rewriter);
+      }
+    };
+
+    auto pattern =
+        std::make_unique<FnPattern>(typeConverter, patterns.getContext());
+    pattern->implFn = implFn;
+    patterns.add(std::move(pattern));
+    return *this;
+  }
```

Comment:
This part doesn't seem to be used, thus also not tested. Maybe it'd be a good idea to add this upstream? Seems like it's also useful outside of CIRCT.

Reply:
Ah I forgot to migrate one of the patterns that actually needs the `adaptor`. Added that. Good idea about upstreaming this!

Reply:
See slide 19 of https://llvm.org/devmtg/2024-10/slides/techtalk/Springer-Pattern-Based-IR-Rewriting-in-MLIR.pdf

Reply:
Hmm yeah, the `RewritePatternSet` uses `OpRewritePattern`. But it's still used to collect the `OpConversionPattern`s in a dialect conversion :thinking:. This new pattern set just factors out creating a subclass of `OpConversionPattern` into a helper function, but the rest is still the same as any other dialect conversion.

Reply:
Yes, that's part of what's meant with "API design suggests that Conversion/RewritePattern are compatible" in the slides. What I was suggesting is to change all `convert` functions in the pass to have the adaptor argument even if it is not used (or change the `add` function in `ConversionPatternSet` that currently delegates to `RewritePatternSet` to also create a `OpConversionPattern` instead)

Reply:
Ah I see! Sorry, slow brain today. Let me change those over to use the adaptor :+1:.